website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md

```diff
@@ -0,0 +1,579 @@
+  rest_api_id = "${aws_api_gateway_rest_api.example.id}"
+  parent_id   = "${aws_api_gateway_rest_api.example.root_resource_id}"
+  path_part   = "{proxy+}"
+}
+
+resource "aws_api_gateway_method" "proxy" {
+  rest_api_id   = "${aws_api_gateway_rest_api.example.id}"
+  resource_id   = "${aws_api_gateway_resource.proxy.id}"
+  http_method   = "ANY"
+  authorization = "NONE"
+}
+```
+
+The special `path_part` value `"{proxy+}"` activates proxy behavior, which
+means that this resource will match _any_ request path. Similarly, the
+`aws_api_gateway_method` block uses a `http_method` of `"ANY"`, which allows
+any request method to be used. Taken together, this means that all incoming
+requests will match this resource.
+
+Each method on an API gateway resource has an _integration_ which specifies
+where incoming requests are routed. Add the following configuration to specify
+that requests to this method should be sent to the Lambda function defined
+earlier:
+
+```hcl
+resource "aws_api_gateway_integration" "lambda" {
+  rest_api_id = "${aws_api_gateway_rest_api.example.id}"
+  resource_id = "${aws_api_gateway_method.proxy.resource_id}"
+  http_method = "${aws_api_gateway_method.proxy.http_method}"
+
+  integration_http_method = "POST"
+  type                    = "AWS_PROXY"
+  uri                     = "${aws_lambda_function.main.invoke_arn}"
+}
+```
+
+The `AWS_PROXY` integration type causes API gateway to call into the API of
+another AWS service. In this case, it will call the AWS Lambda API to create
+an "invocation" of the Lambda function.
+
+Unfortunately the proxy resource cannot match an _empty_ path at the root of
+the API. To handle that, a similar configuration must be applied to the
+_root resource_ that is built in to the REST API object:
+
+```hcl
+resource "aws_api_gateway_method" "proxy_root" {
+  rest_api_id   = "${aws_api_gateway_rest_api.example.id}"
+  resource_id   = "${aws_api_gateway_rest_api.example.root_resource_id}"
+  http_method   = "ANY"
+  authorization = "NONE"
+}
+
+resource "aws_api_gateway_integration" "lambda_root" {
+  rest_api_id = "${aws_api_gateway_rest_api.example.id}"
+  resource_id = "${aws_api_gateway_method.proxy_root.resource_id}"
+  http_method = "${aws_api_gateway_method.proxy_root.http_method}"
+
+  integration_http_method = "POST"
+  type                    = "AWS_PROXY"
+  uri                     = "${aws_lambda_function.main.invoke_arn}"
+}
+```
+
+Finally, configure the Lambda function to allow this API Gateway API to make
+requests to it by appending the following to the `lambda.tf` file created in
+an earlier step:
+
+```hcl
+resource "aws_lambda_permission" "apigw" {
+  statement_id  = "AllowAPIGatewayInvoke"
+  action        = "lambda:InvokeFunction"
+  function_name = "${aws_lambda_function.example.arn}"
+  principal     = "apigateway.amazonaws.com"
+
+  source_arn = "arn:aws:execute-api:us-east-1:${data.aws_caller_identity.current.account_id}:${aws_api_gateway_rest_api.example.id}/*/*/*"
```

Comment:
Yeah, the sequencing of the guide here was the reason I ended up not using that here. (Actually, I only remembered about `execution_arn` after I'd passed this point in the guide myself!)

I did consider restructuring the guide to do that step first but I wanted to complete the setup of the API before context-switching back to Lambda, because it feels confusing to keep switching back and forth between the two services.

Reply:
Actually now I re-read I remember that I already did re-work this slightly since my original draft and so the permission setup is already interleaved in there. It would actually make it _less_ context-switchy if I reordered it to do the deployment earlier.

That would then remove the need for the "test it in the API Gateway console" step, which is probably for the best since it's annoying to explain how to get there in the UI.